File: go/stego.go
package main

import (
	"bytes"
	"compress/flate"
	"errors"
	"fmt"
	"io"
	"strings"
	"unicode/utf8"
)

// HideOptions controls how a secret is packed.
type HideOptions struct {
	Password   string // empty means "do not encrypt"
	Codec      Codec
	Iterations int
	NoCompress bool
}

// Report describes what was produced or found, for display purposes.
type Report struct {
	Encrypted  bool
	Compressed bool
	Codec      Codec
	Iterations int
	SecretSize int // plaintext bytes (0 when unknown, i.e. scanning a sealed payload)
	Container  int // container bytes on the wire
	Carriers   int // invisible characters
}

// Hide packs secret into an invisible payload and injects it into cover.
func Hide(cover, secret string, opts HideOptions) (string, Report, error) {
	if secret == "" {
		return "", Report{}, errors.New("the secret is empty — there is nothing to hide")
	}
	if opts.Codec == 0 {
		opts.Codec = CodecB4
	}
	if opts.Iterations == 0 {
		opts.Iterations = defaultIterations
	}

	body := []byte(secret)
	rep := Report{Codec: opts.Codec, SecretSize: len(body)}

	// Compression happens before encryption: it shrinks the payload — often by
	// 3-5x on natural language, which matters a lot when every byte costs four
	// invisible characters — and it is safe here because the attacker never
	// gets to inject chosen plaintext into the secret.
	if !opts.NoCompress {
		if squeezed, err := deflate(body); err == nil && len(squeezed) < len(body) {
			body = squeezed
			rep.Compressed = true
		}
	}

	c := &container{}
	if rep.Compressed {
		c.flags |= flagCompressed
	}

	var payload []byte

	if opts.Password == "" {
		inner, err := packInner(body, false)
		if err != nil {
			return "", Report{}, err
		}
		payload = inner
		c.payload = payload
	} else {
		c.flags |= flagEncrypted
		c.kdf = kdfPBKDF2SHA256
		c.iterations = uint32(opts.Iterations)
		c.salt = make([]byte, saltLen)
		c.nonce = make([]byte, nonceLen)
		if err := randBytes(c.salt); err != nil {
			return "", Report{}, err
		}
		if err := randBytes(c.nonce); err != nil {
			return "", Report{}, err
		}

		inner, err := packInner(body, true)
		if err != nil {
			return "", Report{}, err
		}

		key, err := deriveKey(opts.Password, c.salt, opts.Iterations)
		if err != nil {
			return "", Report{}, err
		}
		defer wipe(key)

		// The header is authenticated, so the declared length must be the
		// final sealed length — computed up front, before sealing.
		aad := c.header(len(inner) + gcmOverhead)
		payload, err = seal(key, c.nonce, inner, aad)
		if err != nil {
			return "", Report{}, err
		}
		c.payload = payload

		rep.Encrypted = true
		rep.Iterations = opts.Iterations
	}

	raw := c.marshal()
	invisible := Encode(raw, opts.Codec)

	rep.Container = len(raw)
	rep.Carriers = utf8.RuneCountInString(invisible)

	return Inject(cover, invisible), rep, nil
}

// Scan recovers and identifies a payload without decrypting it. The returned
// container is still sealed; pass it to Unseal with a password.
func Scan(text string) (*container, []byte, Codec, Report, error) {
	for _, codec := range []Codec{CodecB4, CodecB2} {
		raw := Decode(text, codec)
		if len(raw) == 0 {
			continue
		}
		c, header, err := parseContainer(raw)
		if err != nil {
			// Only a magic mismatch means "wrong codec, keep looking"; a real
			// structural problem should be reported to the user as-is.
			if errors.Is(err, errBadMagic) || errors.Is(err, errNoPayload) {
				continue
			}
			return nil, nil, codec, Report{}, err
		}
		rep := Report{
			Encrypted:  c.encrypted(),
			Compressed: c.compressed(),
			Codec:      codec,
			Iterations: int(c.iterations),
			Container:  len(header) + len(c.payload),
			Carriers:   CountCarriers(text),
		}
		return c, header, codec, rep, nil
	}

	// Nothing modern matched. Fall back to the v1/v2 layouts.
	if secret, ok := decodeLegacy(text); ok {
		return nil, nil, CodecB2, Report{
			Codec:      CodecB2,
			SecretSize: len(secret),
			Carriers:   CountCarriers(text),
		}, errLegacyPlain{secret}
	}

	if CountCarriers(text) == 0 {
		return nil, nil, 0, Report{}, errNoPayload
	}
	return nil, nil, 0, Report{}, errBadMagic
}

// Unseal turns a scanned container into the original secret. password is
// ignored for unencrypted payloads.
func Unseal(c *container, header []byte, password string) (string, error) {
	var inner []byte

	if c.encrypted() {
		if password == "" {
			return "", errors.New("this payload is encrypted and needs a password")
		}
		key, err := deriveKey(password, c.salt, int(c.iterations))
		if err != nil {
			return "", err
		}
		defer wipe(key)

		inner, err = open(key, c.nonce, c.payload, header)
		if err != nil {
			return "", err
		}
	} else {
		inner = c.payload
	}

	body, err := unpackInner(inner)
	if err != nil {
		return "", err
	}

	if c.compressed() {
		body, err = inflate(body)
		if err != nil {
			return "", fmt.Errorf("the payload could not be decompressed: %w", err)
		}
	}
	if !utf8.Valid(body) {
		return "", errors.New("the recovered data is not valid text")
	}
	return string(body), nil
}

// Reveal is the one-shot path: scan, ask for a password if needed, unseal.
// askPassword may be nil for payloads that are known to be unencrypted.
func Reveal(text string, askPassword func() (string, bool)) (string, Report, error) {
	c, header, _, rep, err := Scan(text)
	if err != nil {
		var legacy errLegacyPlain
		if errors.As(err, &legacy) {
			return legacy.secret, rep, nil
		}
		return "", rep, err
	}

	password := ""
	if c.encrypted() {
		if askPassword == nil {
			return "", rep, errors.New("this payload is encrypted and needs a password")
		}
		pw, ok := askPassword()
		if !ok {
			return "", rep, errAborted
		}
		password = pw
	}

	secret, err := Unseal(c, header, password)
	if err != nil {
		return "", rep, err
	}
	rep.SecretSize = len(secret)
	return secret, rep, nil
}

var errAborted = errors.New("operation cancelled")

// errLegacyPlain carries a secret recovered from a pre-v3 payload, which has
// no container to hand back.
type errLegacyPlain struct{ secret string }

func (errLegacyPlain) Error() string { return "legacy payload" }

// decodeLegacy reads the v1 (STEGO_V1:) and v2 (STEGO_RAW:) layouts so older
// messages keep working. v2's encrypted form (STEGO_ENC:) is not supported —
// it derived its key with a bare SHA-256 and is deliberately not carried
// forward.
func decodeLegacy(text string) (string, bool) {
	raw := Decode(text, CodecB2)
	if len(raw) == 0 {
		return "", false
	}
	for _, prefix := range []string{"STEGO_V1:", "STEGO_RAW:"} {
		if bytes.HasPrefix(raw, []byte(prefix)) {
			body := raw[len(prefix):]
			if utf8.Valid(body) {
				return string(body), true
			}
		}
	}
	return "", false
}

// LegacyEncrypted reports whether text holds a v2 STEGO_ENC: payload, so the
// user can be told exactly why it is being refused instead of getting a
// generic "unknown header".
func LegacyEncrypted(text string) bool {
	raw := Decode(text, CodecB2)
	return bytes.HasPrefix(raw, []byte("STEGO_ENC:"))
}

func deflate(data []byte) ([]byte, error) {
	var buf bytes.Buffer
	w, err := flate.NewWriter(&buf, flate.BestCompression)
	if err != nil {
		return nil, err
	}
	if _, err := w.Write(data); err != nil {
		return nil, err
	}
	if err := w.Close(); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

func inflate(data []byte) ([]byte, error) {
	r := flate.NewReader(bytes.NewReader(data))
	defer r.Close()
	// Bound the output so a malicious payload cannot be a decompression bomb.
	out, err := io.ReadAll(io.LimitReader(r, maxPayload+1))
	if err != nil {
		return nil, err
	}
	if len(out) > maxPayload {
		return nil, errors.New("decompressed data exceeds the size limit")
	}
	return out, nil
}

// Describe renders a Report as a short human-readable summary.
func (r Report) Describe() string {
	var parts []string
	if r.Encrypted {
		parts = append(parts, fmt.Sprintf("AES-256-GCM, PBKDF2-SHA256 x%d", r.Iterations))
	} else {
		parts = append(parts, "not encrypted")
	}
	if r.Compressed {
		parts = append(parts, "DEFLATE")
	}
	parts = append(parts, r.Codec.String())
	return strings.Join(parts, " | ")
}

File: android/java/ir/sayeh/stego/Stego.java
package ir.sayeh.stego;

import java.io.ByteArrayOutputStream;
import java.nio.charset.CharacterCodingException;
import java.nio.charset.CharsetDecoder;
import java.nio.charset.CodingErrorAction;
import java.nio.charset.StandardCharsets;
import java.nio.ByteBuffer;
import java.security.GeneralSecurityException;
import java.util.zip.DataFormatException;
import java.util.zip.Deflater;
import java.util.zip.Inflater;

import javax.crypto.AEADBadTagException;

/**
 * High-level hide/reveal API. Pure java.* and javax.* on purpose — no Android
 * imports — so the exact code that ships in the app can also be run and
 * verified on a desktop JVM (see SelfTest).
 */
public final class Stego {

    /** Options for hiding. */
    public static final class Options {
        public String password = "";     // empty means "do not encrypt"
        public int codec = ZwCodec.B4;
        public int iterations = Crypto.DEFAULT_ITERATIONS;
        public boolean noCompress = false;
    }

    /** What was produced or found, for display. */
    public static final class Report {
        public boolean encrypted;
        public boolean compressed;
        public int codec = ZwCodec.B4;
        public int iterations;
        public int secretSize;
        public int containerSize;
        public int carriers;
        public boolean legacy;

        public String describe() {
            StringBuilder sb = new StringBuilder();
            sb.append(encrypted
                    ? "AES-256-GCM, PBKDF2-SHA256 x" + iterations
                    : "not encrypted");
            if (compressed) {
                sb.append(" | DEFLATE");
            }
            sb.append(" | ").append(ZwCodec.name(codec));
            if (legacy) {
                sb.append(" | legacy v1/v2");
            }
            return sb.toString();
        }
    }

    /** A scan result: the sealed container plus what we can say about it. */
    public static final class Scanned {
        public Format.Container container; // null for a legacy payload
        public String legacySecret;        // non-null only for legacy payloads
        public Report report = new Report();

        public boolean needsPassword() {
            return container != null && container.isEncrypted();
        }
    }

    /** Hiding produces the stego text plus a report. */
    public static final class Hidden {
        public String text;
        public Report report = new Report();
    }

    private Stego() {
    }

    // -----------------------------------------------------------------------
    // Hide
    // -----------------------------------------------------------------------

    public static Hidden hide(String cover, String secret, Options opts)
            throws StegoException {

        if (secret == null || secret.isEmpty()) {
            throw new StegoException(StegoException.Kind.EMPTY_SECRET,
                    "The secret is empty — there is nothing to hide.");
        }
        if (cover == null) {
            cover = "";
        }

        byte[] body = secret.getBytes(StandardCharsets.UTF_8);

        Hidden result = new Hidden();
        result.report.codec = opts.codec;
        result.report.secretSize = body.length;

        // Compress before encrypting: it often shrinks natural language by
        // 3-5x, which matters when every byte costs four invisible characters.
        // Safe here because an attacker cannot inject chosen plaintext into
        // the secret, which is what makes compression-then-encryption risky
        // in protocols like TLS.
        if (!opts.noCompress) {
            byte[] squeezed = deflate(body);
            if (squeezed.length < body.length) {
                body = squeezed;
                result.report.compressed = true;
            }
        }

        Format.Container c = new Format.Container();
        if (result.report.compressed) {
            c.flags |= Format.FLAG_COMPRESSED;
        }

        try {
            if (opts.password == null || opts.password.isEmpty()) {
                c.payload = Format.packInner(body, false);
            } else {
                c.flags |= Format.FLAG_ENCRYPTED;
                c.kdf = Format.KDF_PBKDF2_SHA256;
                c.iterations = opts.iterations;
                c.salt = Crypto.random(Crypto.SALT_LEN);
                c.nonce = Crypto.random(Crypto.NONCE_LEN);

                byte[] inner = Format.packInner(body, true);
                byte[] key = Crypto.deriveKey(opts.password, c.salt, c.iterations);
                try {
                    byte[] aad = c.buildHeader(inner.length + Crypto.GCM_OVERHEAD);
                    c.payload = Crypto.seal(key, c.nonce, inner, aad);
                } finally {
                    Crypto.wipe(key);
                }

                result.report.encrypted = true;
                result.report.iterations = opts.iterations;
            }
        } catch (GeneralSecurityException e) {
            throw new StegoException(StegoException.Kind.CORRUPT,
                    "Encryption failed: " + e.getMessage());
        }

        byte[] raw = c.marshal();
        String invisible = ZwCodec.encode(raw, opts.codec);

        result.report.containerSize = raw.length;
        result.report.carriers = invisible.length();
        result.text = ZwCodec.inject(cover, invisible);
        return result;
    }

    // -----------------------------------------------------------------------
    // Scan / reveal
    // -----------------------------------------------------------------------

    /** Recovers and identifies a payload without decrypting it. */
    public static Scanned scan(String text) throws StegoException {
        if (text == null) {
            text = "";
        }

        StegoException structural = null;

        for (int codec : new int[]{ZwCodec.B4, ZwCodec.B2}) {
            byte[] raw = ZwCodec.decode(text, codec);
            if (raw.length == 0) {
                continue;
            }
            try {
                Format.Container c = Format.parse(raw);

                Scanned s = new Scanned();
                s.container = c;
                s.report.encrypted = c.isEncrypted();
                s.report.compressed = c.isCompressed();
                s.report.codec = codec;
                s.report.iterations = c.iterations;
                s.report.containerSize = c.header.length + c.payload.length;
                s.report.carriers = ZwCodec.countCarriers(text);
                return s;
            } catch (StegoException e) {
                // Only a magic mismatch means "wrong codec, keep looking".
                if (e.kind == StegoException.Kind.BAD_MAGIC
                        || e.kind == StegoException.Kind.NO_PAYLOAD) {
                    continue;
                }
                structural = e;
            }
        }
        if (structural != null) {
            throw structural;
        }

        String legacy = decodeLegacy(text);
        if (legacy != null) {
            Scanned s = new Scanned();
            s.legacySecret = legacy;
            s.report.legacy = true;
            s.report.codec = ZwCodec.B2;
            s.report.secretSize = legacy.getBytes(StandardCharsets.UTF_8).length;
            s.report.carriers = ZwCodec.countCarriers(text);
            return s;
        }

        if (ZwCodec.countCarriers(text) == 0) {
            throw StegoException.noPayload();
        }
        throw StegoException.badMagic();
    }

    /** Turns a scanned container into the original secret. */
    public static String unseal(Format.Container c, String password) throws StegoException {
        byte[] inner;

        if (c.isEncrypted()) {
            if (password == null || password.isEmpty()) {
                throw new StegoException(StegoException.Kind.NEEDS_PASSWORD,
                        "This payload is encrypted and needs a password.");
            }
            byte[] key = null;
            try {
                key = Crypto.deriveKey(password, c.salt, c.iterations);
                inner = Crypto.open(key, c.nonce, c.payload, c.header);
            } catch (AEADBadTagException e) {
                throw StegoException.wrongPassword();
            } catch (GeneralSecurityException e) {
                // Every other failure mode is also "it did not authenticate";
                // reporting them separately would leak which part went wrong.
                throw StegoException.wrongPassword();
            } finally {
                Crypto.wipe(key);
            }
        } else {
            inner = c.payload;
        }

        byte[] body = Format.unpackInner(inner);

        if (c.isCompressed()) {
            try {
                body = inflate(body);
            } catch (DataFormatException e) {
                throw new StegoException(StegoException.Kind.CORRUPT,
                        "The payload could not be decompressed.");
            }
        }
        return decodeUtf8(body);
    }

    /** Scan and unseal in one step. */
    public static String reveal(String text, String password) throws StegoException {
        Scanned s = scan(text);
        if (s.legacySecret != null) {
            return s.legacySecret;
        }
        return unseal(s.container, password);
    }

    /**
     * Reads the v1 ({@code STEGO_V1:}) and v2 ({@code STEGO_RAW:}) layouts so
     * older messages keep working. v2's encrypted form is not carried forward:
     * it derived its key with a bare SHA-256.
     */
    public static String decodeLegacy(String text) {
        byte[] raw = ZwCodec.decode(text, ZwCodec.B2);
        if (raw.length == 0) {
            return null;
        }
        for (String prefix : new String[]{"STEGO_V1:", "STEGO_RAW:"}) {
            byte[] p = prefix.getBytes(StandardCharsets.US_ASCII);
            if (Format.startsWith(raw, p)) {
                byte[] body = java.util.Arrays.copyOfRange(raw, p.length, raw.length);
                try {
                    return decodeUtf8(body);
                } catch (StegoException e) {
                    return null;
                }
            }
        }
        return null;
    }

    /** True when the text carries a v2 {@code STEGO_ENC:} payload. */
    public static boolean isLegacyEncrypted(String text) {
        byte[] raw = ZwCodec.decode(text, ZwCodec.B2);
        return Format.startsWith(raw, "STEGO_ENC:".getBytes(StandardCharsets.US_ASCII));
    }

    // -----------------------------------------------------------------------
    // Helpers
    // -----------------------------------------------------------------------

    private static String decodeUtf8(byte[] body) throws StegoException {
        CharsetDecoder decoder = StandardCharsets.UTF_8.newDecoder()
                .onMalformedInput(CodingErrorAction.REPORT)
                .onUnmappableCharacter(CodingErrorAction.REPORT);
        try {
            return decoder.decode(ByteBuffer.wrap(body)).toString();
        } catch (CharacterCodingException e) {
            throw new StegoException(StegoException.Kind.CORRUPT,
                    "The recovered data is not valid text.");
        }
    }

    static byte[] deflate(byte[] data) {
        Deflater d = new Deflater(Deflater.BEST_COMPRESSION, true); // nowrap: raw DEFLATE
        try {
            d.setInput(data);
            d.finish();
            ByteArrayOutputStream out = new ByteArrayOutputStream(data.length);
            byte[] buf = new byte[4096];
            while (!d.finished()) {
                out.write(buf, 0, d.deflate(buf));
            }
            return out.toByteArray();
        } finally {
            d.end();
        }
    }

    static byte[] inflate(byte[] data) throws DataFormatException, StegoException {
        Inflater inf = new Inflater(true);
        try {
            inf.setInput(data);
            ByteArrayOutputStream out = new ByteArrayOutputStream(data.length * 4);
            byte[] buf = new byte[4096];
            while (!inf.finished()) {
                int n = inf.inflate(buf);
                if (n == 0 && (inf.needsInput() || inf.needsDictionary())) {
                    break; // truncated stream
                }
                out.write(buf, 0, n);
                // Bound the output so a hostile payload cannot be a zip bomb.
                if (out.size() > Format.MAX_PAYLOAD) {
                    throw new StegoException(StegoException.Kind.CORRUPT,
                            "Decompressed data exceeds the size limit.");
                }
            }
            return out.toByteArray();
        } finally {
            inf.end();
        }
    }
}
